### PR TITLE
8295060: Port PrintDeoptimizationDetails to UL

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi!

This PR ports PrintDeoptimizationDetails to UL by mapping its output to debug level with tag deoptimization.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295060](https://bugs.openjdk.org/browse/JDK-8295060): Port PrintDeoptimizationDetails to UL

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [5b9f53d3](https://git.openjdk.org/jdk/pull/10645/files/5b9f53d33291c3c23ccbc8aceb1662a0d25bc0ea)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10645/head:pull/10645` \
`$ git checkout pull/10645`

Update a local copy of the PR: \
`$ git checkout pull/10645` \
`$ git pull https://git.openjdk.org/jdk pull/10645/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10645`

View PR using the GUI difftool: \
`$ git pr show -t 10645`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10645.diff">https://git.openjdk.org/jdk/pull/10645.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/vframe.hpp [hunk 1]
```diff
@@ -110,6 +110,7 @@ class vframe: public ResourceObj {
   // printing operations
   virtual void print_value() const;
   virtual void print();
+  void print_on(outputStream* st) const override;
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vframe.hpp:113]: Unclear if this should also be virtual.

File: src/hotspot/share/runtime/vframeArray.cpp [hunk 2]
```diff
@@ -323,19 +324,32 @@ void vframeArrayElement::unpack_on_stack(int caller_actual_parameters,
...
-
   // Unpack expression stack
   // If this is an intermediate frame (i.e. not top frame) then this
   // only unpacks the part of the expression stack not used by callee
   // as parameters. The callee parameters are unpacked as part of the
   // callee locals.
   int i;
+
+#ifndef PRODUCT
+  auto log_deopt = [](int i, intptr_t* addr) {
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/runtime/vframeArray.cpp:338]: ?? I only see one usage in the switch statement so don't understand why this is not inline as normal logging code would be.

File: src/hotspot/share/runtime/vframeArray.cpp [hunk 3]
```diff
@@ -372,13 +375,23 @@ void vframeArrayElement::unpack_on_stack(int caller_actual_parameters,
         ShouldNotReachHere();
     }
   }
-
 #ifndef PRODUCT
-  if (PrintDeoptimizationDetails) {
-    tty->print_cr("Locals size: %d", locals()->size());
-  }
+  log_debug(deoptimization)("Locals size: %d", locals()->size());
+  auto log_it = [](int i, intptr_t* addr) {
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vframeArray.cpp:380]: Again I don't see why this can't just be inline

File: test/hotspot/gtest/logging/test_logStream.cpp [hunk 4]
```diff
@@ -155,4 +155,3 @@ TEST_VM_F(LogStreamTest, autoflush_on_destruction) {
   }
   EXPECT_TRUE(file_contains_substring(TestLogFileName, "ABCD\n"));
 }
-
```
Comment (dholmes-ora) [inline test/hotspot/gtest/logging/test_logStream.cpp:158]: Inadvertent removal?

File: test/hotspot/jtreg/compiler/uncommontrap/TestDeoptOOM.java [hunk 5]
```diff
@@ -42,7 +42,7 @@
  *      -XX:CompileCommand=exclude,compiler.uncommontrap.TestDeoptOOM::main
  *      -XX:CompileCommand=exclude,compiler.uncommontrap.TestDeoptOOM::m9_1
  *      -XX:+UnlockDiagnosticVMOptions
- *      -XX:+UseZGC -XX:+LogCompilation -XX:+PrintDeoptimizationDetails -XX:+TraceDeoptimization -XX:+Verbose
+ *      -XX:+UseZGC -XX:+LogCompilation -XX:+TraceDeoptimization -XX:+Verbose
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/compiler/uncommontrap/TestDeoptOOM.java:45]: Why isn't this enabling deoptimization logging?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Johan,

I have a few comments/suggestions about some of this.

Thanks.

File: src/hotspot/share/runtime/vframeArray.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/runtime/vframeArray.cpp:338]: I should've said "crosses initialization error", not "crossing."

Check out this SO question: https://stackoverflow.com/questions/11578936/getting-a-bunch-of-crosses-initialization-error

So `LogTarget(Debug, deoptimization) lt;` is the error here. I *think* that we can inline it if we introduce a surrounding  scope, is that preferable to you?

So:

```
case T_OBJECT:
*addr = value->get_int(T_OBJECT);
{ // Scope off LogTarget
    LogTarget(Debug, deoptimization) lt;
    if (lt.is_enabled()) {
      LogStream ls(lt);
      ls.print(" - Reconstructed expression %d (OBJECT): ", i);
      oop o = cast_to_oop((address)(*addr));
      if (o == NULL) {
        ls.print_cr("NULL");
      } else {
        ResourceMark rm;
        ls.print_raw_cr(o->klass()->name()->as_C_string());
      }
    }
}
```

File: src/hotspot/share/runtime/vframe.hpp (hunk 1 above)
Comment (jdksjolen) [reply src/hotspot/share/runtime/vframe.hpp:113]: This is virtual, taken from `ResourceObj`. The `override` indicates this, I'm basing this on: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-override

File: src/hotspot/share/runtime/vframeArray.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/runtime/vframeArray.cpp:338]: Yes adding the scope is fine and preferable. Thanks.

File: src/hotspot/share/runtime/vframe.hpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/vframe.hpp:113]: Thanks - missed that.

File: src/hotspot/share/runtime/vframe.cpp [hunk 6]
```diff
@@ -678,7 +678,10 @@ javaVFrame* vframeStreamCommon::asJavaVFrame() {
 
 #ifndef PRODUCT
 void vframe::print() {
-  if (WizardMode) _fr.print_value_on(tty,NULL);
+  if (WizardMode) print_on(tty);
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/runtime/vframe.cpp]: This sounds OK to me.

**Review by jdksjolen — COMMENTED** [review]
Finished up the comments that David left, thanks for the review!

**Comment by jdksjolen** [comment]
Okay, Github's UI tricked me a bit here. I thought my comments would be applied as replies to David's comments, but they're separate...

File: test/hotspot/jtreg/compiler/uncommontrap/TestDeoptOOM.java [hunk 7]
```diff
@@ -42,7 +42,7 @@
  *      -XX:CompileCommand=exclude,compiler.uncommontrap.TestDeoptOOM::main
  *      -XX:CompileCommand=exclude,compiler.uncommontrap.TestDeoptOOM::m9_1
  *      -XX:+UnlockDiagnosticVMOptions
- *      -XX:+UseZGC -XX:+LogCompilation -XX:+PrintDeoptimizationDetails -XX:+TraceDeoptimization -XX:+Verbose
+ *      -XX:+UseZGC -XX:+LogCompilation -Xlog:disable -Xlog:deoptimization=trace -XX:+TraceDeoptimization -XX:+Verbose
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/uncommontrap/TestDeoptOOM.java:45]: Why you need `-Xlog:disable` ?

**Review by vnkozlov — COMMENTED** [review]
Nice work. Can you show an example of new output vs old?

File: test/hotspot/jtreg/compiler/uncommontrap/TestDeoptOOM.java (hunk 7 above)
Comment (jdksjolen) [reply test/hotspot/jtreg/compiler/uncommontrap/TestDeoptOOM.java]: Short version: Mistake on my part, it shouldn't be there.

Long version: `-Xlog:disable` disables the default output to stdout, this is important when you want your logs to *not* go to stdout at all. Of course, we want our logs to go to stdout, so this is unnecessary here.

**Comment by jdksjolen** [comment]
> Nice work. Can you show an example of new output vs old?

New: `-Xlog:deoptimization=trace:deopt.txt`

```
[1,923s][debug][deoptimization] Expressions size: 1
[1,923s][debug][deoptimization]  - Reconstructed expression 0 (OBJECT): NULL
[1,923s][debug][deoptimization] Locals size: 3
[1,923s][debug][deoptimization]  - Reconstructed local 0 (OBJECT): java/lang/String
[1,923s][debug][deoptimization]  - Reconstructed local 1 (OBJECT): NULL
[1,924s][debug][deoptimization] [1. Interpreted Frame]
[1,924s][debug][deoptimization] Interpreted frame (sp=0x00007fd07e205548 unextended sp=0x00007fd07e205548, fp=0x00007fd07e205598, real_fp=0x00007fd07e205598, pc=0x00007fd0704506a0)
[1,924s][debug][deoptimization] ~deoptimization entry points  [0x00007fd0704506a0, 0x00007fd070453b38]  13464 bytes
[1,924s][debug][deoptimization]      BufferBlob (0x00007fd070431510) used for Interpreter
[1,924s][debug][deoptimization]  - local  [0x000000011d00d0e8]; #0
[1,924s][debug][deoptimization]  - local  [0x0000000000000000]; #1
[1,924s][debug][deoptimization]  - local  [0x0000000000000000]; #2
[1,924s][debug][deoptimization]  - stack  [0x0000000000000000]; #0
[1,924s][debug][deoptimization]  - monitor[0x00007fd07e205550]
[1,924s][debug][deoptimization]  - bcp    [0x00007fd02800b230]; @8
[1,924s][debug][deoptimization]  - locals [0x00007fd07e2055b8]
[1,924s][debug][deoptimization]  - method [0x00007fd02800b298]; virtual jboolean java.lang.String.equals(jobject)
[1,924s][debug][deoptimization] 
```

Old: 

``` `-XX:+WizardMode -XX:+Verbose -XX:+PrintDeoptimizationDetails`
Expressions size: 1
 - Reconstructed expression 0 (OBJECT): NULL
Locals size: 3
 - Reconstructed local 0 (OBJECT): java/lang/String
 - Reconstructed local 1 (OBJECT): NULL
[1. Interpreted Frame]
Interpreted frame (sp=0x00007fc830059548 unextended sp=0x00007fc830059548, fp=0x00007fc830059598, real_fp=0x00007fc830059598, pc=0x00007fc8204506a0)
~deoptimization entry points  [0x00007fc8204506a0, 0x00007fc820453b38]  13464 bytes
     BufferBlob (0x00007fc820431510) used for Interpreter
Trying to load: /home/johan/jdk/build/linux-x64-slowdebug/jdk/lib/server/libhsdis-amd64.so
Trying to load: /home/johan/jdk/build/linux-x64-slowdebug/jdk/lib/server/hsdis-amd64.so
Trying to load: /home/johan/jdk/build/linux-x64-slowdebug/jdk/lib/hsdis-amd64.so
Trying to load: hsdis-amd64.so
[1,857s][warning][os] Loading hsdis library failed
Could not load hsdis-amd64.so; hsdis-amd64.so: cannot open shared object file: No such file or directory; PrintAssembly defaults to abstract disassembly.
[MachCode]
// Snip! Huge amount of logs here
[/MachCode]
```

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 8]
```diff
@@ -296,32 +296,31 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
                                                      exec_mode, realloc_failures);
       deoptimized_objects = deoptimized_objects || relocked;
 #ifndef PRODUCT
-      if (PrintDeoptimizationDetails) {
-        ResourceMark rm;
-        stringStream st;
+      LogMessage(deoptimization) lm;
```
Comment (tstuefe) [inline src/hotspot/share/runtime/deoptimization.cpp:299]: Drive-by: Does this not already incur costs? Does LogMessage not contain an internal buffer it allocates?

File: src/hotspot/share/runtime/vframe.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/runtime/vframe.hpp:113]: I don't know if we've added 'override' to our source base yet.

**Review by coleenp — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/runtime/vframeArray.cpp [hunk 9]
```diff
@@ -446,23 +444,24 @@ void vframeArrayElement::unpack_on_stack(int caller_actual_parameters,
...
+      NonInterleavingLogStream ls(LogLevelType::Debug, lm);
+      ls.print_cr("[%d. Interpreted Frame]", ++unpack_counter);
+      iframe()->print_on(&ls);
+      RegisterMap map(thread,
+                      RegisterMap::UpdateMap::include,
+                      RegisterMap::ProcessFrames::include,
+                      RegisterMap::WalkContinuation::skip);
+      vframe* f = vframe::new_vframe(iframe(), &map, thread);
+      f->print_on(&ls);
+      if (WizardMode && Verbose) method()->print_codes_on(&ls);
```
Comment (coleenp) [inline src/hotspot/share/runtime/vframeArray.cpp:460]: I guess David already commented on this but we're trying to remove WizardMode and Verbose.  Can this be:
if (lm.is_trace()) method->print_codes_on(lm);
?
Also methodHandle doesn't require () because -> turns it into Method*.

**Comment by vnkozlov** [comment]
> Old `-XX:+WizardMode -XX:+Verbose -XX:+PrintDeoptimizationDetails`:

You only need `-XX:+PrintDeoptimizationDetails` to see old output. `-XX:+WizardMode -XX:+Verbose` will mess it up.

I don't see next output in your example after `- method ` last line:
```
        {method} {0x0000000800009b10} 'equals' '(Ljava/lang/Object;)Z' in 'java/lang/String'
        bci:    8
        locals:
           0 "true"{0x00000007ff84c3a0} <0x00000007ff84c3a0>
           1 NULL <0x0000000000000000>
           2 0 (int) 0.000000 (float) 0 (hex)
        expressions:
           0 NULL <0x0000000000000000>
```

**Comment by vnkozlov** [comment]
I see in new output next lines repeated instead of missing lines from old output:
```
0.703s][debug][deoptimization] Interpreted frame (sp=0x00007fb5bbf165b8 unextended sp=0x00007fb5bbf165b8, fp=0x00007fb5bbf16608, real_fp=0x00007fb5bbf16608, pc=0x00007fb5a43c26a0)
[0.703s][debug][deoptimization] ~deoptimization entry points  [0x00007fb5a43c26a0, 0x00007fb5a43c5b38]  13464 bytes
[0.703s][debug][deoptimization]      BufferBlob (0x00007fb5a43a3510) used for Interpreter
```

File: src/hotspot/share/runtime/vframe.cpp [hunk 10]
```diff
@@ -678,10 +678,10 @@ javaVFrame* vframeStreamCommon::asJavaVFrame() {
 
 #ifndef PRODUCT
 void vframe::print() {
-  if (WizardMode) print_on(tty);
+  print_on(tty);
 }
 void vframe::print_on(outputStream* st) const {
-  if (WizardMode) _fr.print_value_on(st,NULL);
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/runtime/vframe.cpp]: There's a couple other printing function that call vframe::print() so I think that should keep WizardMode on for now until the callers become UL converted.
I don't think the vframe::print_on() version should keep WizardMode since logging calls it with the verbose logging level (guessing this is true, please check!)

[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/runtime/vframe.cpp]: >You don't need the WizardMode guard here and in print_on. UL is supposed replace WizardMode and Verbose so the correct fix would be to elide it from print_on and only log using print_on when the logging level is logically equivalent to WizardMode (as you have done elsewhere).

It was actually in response to what you said :-). I'm OK with either way.

**Comment by jdksjolen** [comment]
I made a new comparison and found this:

- UL adds a line, which comes from this PR: https://github.com/openjdk/jdk/pull/8812
- UL has the precise description of locals missing, looks like this in PrintDeoptimizationDetails:

```
	locals:
	   0 "true"{0x000000011d00cd40} <0x000000011d00cd40>
	   1 NULL <0x0000000000000000>
	   2 0 (int) 0,000000 (float) 0 (hex)
	expressions:
	   0 NULL <0x0000000000000000>
```

No clue why this is missing so far, looking into it.

**Comment by jdksjolen** [comment]
Regarding:

>UL has the precise description of locals missing, looks like this in PrintDeoptimizationDetails:

This is not printed because that info comes from `javaVFrame::print`. `vframe::new_vframe` can return quite a few different `vframe` subclasses, and as such each of these needs to have their `print` be refactored into `print_on + print`.
